### PR TITLE
Docker, docker, docker, …

### LINKED ISSUE #378: Target Docker embeds many dependencies, probably too many...
The latest `roboconf-target-docker` bundle is very heavy (~12MB). Despite this is not a problem itself, I've opened and I've seen that it embeds the SLF4J api (`org.slf4j` package). This cause a warning since the embedded SLF4J cannot find any logger implementation. The package should be imported, as the other bundles do.

Maybe some of the other dependencies could be imported too, instead of being embedded.

The jar is fat, let's start a diet! :tropical_drink:

### LINKED ISSUE #388: Installation of a custom Roboconf plugin
I have written a roboconf plugin to manage Petals ESB service unit. My roboconf application is  deployed on Docker containers. All my container images are generated from the base image `ubuntu:15:04`

The only way to install the Roboconf plugin is to create manually my own images embedding the Robonconf plugin. **As my images are generated by Roboconf, I can't install the Roboconf plugin**. So another way should be available.

Moreover, I think that the production operator has not to install the Robocong plugin in the Roboconf agent. It can be potentially installed into the Roboconf DM that will push the Roboconf plugin in the right Roboconf agent instances (which ones that need it, according to the application graph).

### PR DESCRIPTION
**Pull request by bourretp**
Again, some improvement in the Docker target handler:
- Fixes #388, using the docker.additional.deploy list-of-URL-property
- Fixes #378 by removed some embedded dependencies that can be imported.

### CODE REVIEW

**Comment by bourretp** [comment]
I'll update the site asap (tomorrow).

File: core/roboconf-target-docker/pom.xml [hunk 1]
```diff
@@ -68,6 +68,48 @@
...
+			<groupId>org.slf4j</groupId>
+			<artifactId>slf4j-api</artifactId>
+			<version>1.7.5</version>
+			<scope>provided</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-core</artifactId>
+			<version>2.3.1</version>
```
Comment (vincent-zurczak) [inline core/roboconf-target-docker/pom.xml]: Use ${jackson.stack.version}

File: core/roboconf-target-docker/pom.xml [hunk 2]
```diff
@@ -68,6 +68,48 @@
...
+			<scope>provided</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-databind</artifactId>
```
Comment (vincent-zurczak) [inline core/roboconf-target-docker/pom.xml]: ${jackson.stack.version} (which is 2.3.1)

File: core/roboconf-target-docker/pom.xml [hunk 3]
```diff
@@ -68,6 +68,48 @@
...
+			<version>2.3.0</version>
+			<scope>provided</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+			<artifactId>jackson-annotations</artifactId>
+			<version>2.3.1</version>
```
Comment (vincent-zurczak) [inline core/roboconf-target-docker/pom.xml]: ${jackson.stack.version}

File: core/roboconf-target-docker/pom.xml [hunk 4]
```diff
@@ -68,6 +68,48 @@
...
+			<scope>provided</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.jaxrs</groupId>
+			<artifactId>jackson-jaxrs-json-provider</artifactId>
+			<version>2.1.2</version>
```
Comment (vincent-zurczak) [inline core/roboconf-target-docker/pom.xml]: ${jackson.stack.version}

File: core/roboconf-target-docker/src/test/java/net/roboconf/target/docker/internal/DockerHandler_withPackagesTest.java [hunk 5]
```diff
@@ -403,6 +425,23 @@ private void checkPackageIsInstalled( String packageName ) throws Exception {
...
+	 * Check that a given file is present on the given Docker container.
+	 *
+	 * @param path the path of the file to check.
+	 * @throws TargetException if the container cannot be reached.
+	 */
+	private void checkFileIsPresent( String path ) throws Exception {
+
+		// Execute the package checker command.
+		// As we use pipes, we need to bash -c the whole quoted command.
+		Assert.assertEquals("file '" + path + "' is not present on container " + this.dockerContainerId,
```
Comment (vincent-zurczak) [inline core/roboconf-target-docker/src/test/java/net/roboconf/target/docker/internal/DockerHandler_withPackagesTest.java]: This formatting is ugly. :pill:

File: core/roboconf-target-docker/src/main/java/net/roboconf/target/docker/internal/DockerfileGenerator.java [hunk 6]
```diff
@@ -144,6 +148,13 @@ public File generateDockerfile() throws IOException {
 			out.println("COPY rc.local /etc/");
 			out.println("COPY start.sh /usr/local/roboconf-agent/");
 
+			// Copy additional resources in the Karaf deploy directory if needed.
+			// We use the Docker ADD command that accepts remote URLs.
+			if ( this.deployList != null && !this.deployList.isEmpty() ) {
+				for ( final String deploy : this.deployList )
+					out.println("ADD " + deploy + " /usr/local/roboconf-agent/deploy/");
```
Comment (vincent-zurczak) [inline core/roboconf-target-docker/src/main/java/net/roboconf/target/docker/internal/DockerfileGenerator.java]: I am not sure an URL like file:/... would work with this solution.

**Comment by vincent-zurczak** [comment]
Not bad, but...

* 2 issues should mean 2 commits, one per issue.
* The builds failed.
* Local file URL's should be supported.

File: core/roboconf-target-docker/src/main/java/net/roboconf/target/docker/internal/DockerfileGenerator.java (hunk 6 above)
Comment (bourretp) [inline core/roboconf-target-docker/src/main/java/net/roboconf/target/docker/internal/DockerfileGenerator.java]: You're right, it does not work.
Docker `ADD` command does not accept `file:` URLs, and cannot add local files that are not in the build context (like `COPY`). So, the solution seems to copy all `file:` resources in the context directory, and to `ADD` them after. Problem could be clash of resource names, but that's a corner case.

**Comment by bourretp** [comment]
- I had to handle these issues in a rush, so no time to rewrite the history according to the Great Convention :P
- Yes, the build fails, and it will continue to fail spuriously until we fix all the race conditions everywhere. If you take a look at the build results, 2 builds of 3 pass, and the remaining one's failure in not related to the docker target handler.
- Last point is getting fixed

**Comment by bourretp** [comment]
Fixed "file://" URL... and the build passes (it's magic :D)
Ready to merge as soon as I fix the remaining cosmetics.

File: core/roboconf-target-docker/pom.xml (hunk 4 above)
Comment (bourretp) [inline core/roboconf-target-docker/pom.xml]: ${jackson.stack.version} (2.3.1) should not be compatible with required version 2.1.2.
Keeping this one.

File: core/roboconf-target-docker/pom.xml (hunk 1 above)
Comment (bourretp) [inline core/roboconf-target-docker/pom.xml]: Fixed in 2d44b47

File: core/roboconf-target-docker/pom.xml (hunk 2 above)
Comment (bourretp) [inline core/roboconf-target-docker/pom.xml]: Fixed in 2d44b47

File: core/roboconf-target-docker/pom.xml (hunk 3 above)
Comment (bourretp) [inline core/roboconf-target-docker/pom.xml]: Fixed in 2d44b47

File: core/roboconf-target-docker/pom.xml (hunk 4 above)
Comment (bourretp) [inline core/roboconf-target-docker/pom.xml]: Fixed in 2d44b47

File: core/roboconf-target-docker/src/test/java/net/roboconf/target/docker/internal/DockerHandler_withPackagesTest.java (hunk 5 above)
Comment (bourretp) [inline core/roboconf-target-docker/src/test/java/net/roboconf/target/docker/internal/DockerHandler_withPackagesTest.java]: Fixed in 2d44b47

**Comment by cdeneux** [comment]
Please add support of PAX to manage URL as `mvn:org.ow2.petals/petals-roboconf-plugin/1.0.0-SNAPSHOT`.
For a plugin developped with Maven, `file://` requires an absolute path file that is not portable: `file:///home/cdeneux/.m2/repository/...`, and `http://` requires to deploy manually the Roboconf plugin before to package the Roboconf application

**Comment by vincent-zurczak** [comment]
Please, create an issue for this.
A (closed) pull request is not the best place for this. :)

**Pull request merged**